stdlib/public/core/Slab.swift

```diff
@@ -0,0 +1,479 @@
+// Reduce and Swap
+//===----------------------------------------------------------------------===//
+
+@available(SwiftStdlib 6.1, *)
+extension Slab where Element: ~Copyable {
+  /// Exchanges the values at the specified indices of the vector.
+  ///
+  /// Both parameters must be valid indices of the vector and not
+  /// equal to `endIndex`. Passing the same index as both `i` and `j` has no
+  /// effect.
+  ///
+  /// - Parameters:
+  ///   - i: The index of the first value to swap.
+  ///   - j: The index of the second value to swap.
+  ///
+  /// - Complexity: O(1)
+  @available(SwiftStdlib 6.1, *)
+  @_alwaysEmitIntoClient
+  public mutating func swapAt(
+    _ i: Int,
+    _ j: Int
+  ) {
+    guard i != j, indices.contains(i), indices.contains(j) else {
+      return
+    }
+
+    let ithElement = mutableBuffer.moveElement(from: i)
+    let jthElement = mutableBuffer.moveElement(from: j)
+    mutableBuffer.initializeElement(at: i, to: jthElement)
+    mutableBuffer.initializeElement(at: j, to: ithElement)
+  }
+}
+
+//===----------------------------------------------------------------------===//
+// Unsafe APIs
+//===----------------------------------------------------------------------===//
+
+@available(SwiftStdlib 6.1, *)
+extension Slab where Element: ~Copyable {
+  /// Calls a closure with a pointer to the vector's contiguous storage.
+  ///
+  /// Often, the optimizer can eliminate bounds checks within a vector
+  /// algorithm, but when that fails, invoking the same algorithm on the
+  /// buffer pointer passed into your closure lets you trade safety for speed.
+  ///
+  /// The following example shows how you can iterate over the contents of the
+  /// buffer pointer:
+  ///
+  ///     // "[1, 2, 3, 4, 5]"
+  ///     let numbers = Slab<5, Int> {
+  ///       $0 + 1
+  ///     }
+  ///
+  ///     let sum = numbers.withUnsafeBufferPointer { buffer -> Int in
+  ///         var result = 0
+  ///         for i in stride(from: buffer.startIndex, to: buffer.endIndex, by: 2) {
+  ///             result += buffer[i]
+  ///         }
+  ///         return result
+  ///     }
+  ///     // 'sum' == 9
+  ///
+  /// The pointer passed as an argument to `body` is valid only during the
+  /// execution of `withUnsafeBufferPointer(_:)`. Do not store or return the
+  /// pointer for later use.
+  ///
+  /// - Parameter body: A closure with an `UnsafeBufferPointer` parameter that
+  ///   points to the contiguous storage for the vector. If `body` has a return
+  ///   value, that value is also used as the return value for the
+  ///   `withUnsafeBufferPointer(_:)` method. The pointer argument is valid only
+  ///   for the duration of the method's execution.
+  /// - Returns: The return value, if any, of the `body` closure parameter.
+  @available(SwiftStdlib 6.1, *)
+  @_alwaysEmitIntoClient
+  @_transparent
+  public borrowing func _withUnsafeBufferPointer<Result, E: Error>(
```

Comment:
`Result` should be `Result: ~Copyable` here. Same for the mutable variant below.